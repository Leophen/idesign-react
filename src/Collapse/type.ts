export type CollapseValueType = Array<string | number>;

export interface CollapseProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 当前展开项
   */
  active?: CollapseValueType;
  /**
   * 默认展开项
   */
  defaultActive?: CollapseValueType;
  /**
   * 是否为手风琴模式
   * @default false
   */
  accordion?: boolean;
  /**
   * 全局禁用折叠项
   * @default false
   */
  disabled?: boolean;
  /**
   * 默认全部展开
   * @default false
   */
  expandAll?: boolean;
  /**
   * 自定义图标位置
   * @default left
   */
  iconPlacement?: 'left' | 'right';
  /**
   * 是否隐藏边框
   * @default false
   */
  hideBorder?: boolean;
  /**
   * 无缩进模式
   * @default false
   */
  noIndent?: boolean;
  /**
   * 切换面板时触发，返回变化的值
   */
  onChange?: (value: CollapseValueType) => void;
}

export interface CollapseContextValue {
  inject: (props: CollapseItemProps) => CollapseItemAddProps;
}

export interface CollapseItemProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 折叠项标题
   */
  title?: string;
  /**
   * 折叠项唯一标识
   * @default 索引值
   */
  value?: string | number;
  /**
   * 禁用单折叠项
   * @default false
   */
  disabled?: boolean;
  /**
   * 自定义图标位置
   * @default left
   */
  iconPlacement?: 'left' | 'right';
}

export interface CollapseItemAddProps extends CollapseItemProps {
  /**
   * 当前展开项
   */
  innerActive?: CollapseValueType;
  /**
   * 索引值
   */
  index?: number;
  /**
   * 更新展开项操作
   */
  updateInnerActive?: (value: string | number) => void
}
