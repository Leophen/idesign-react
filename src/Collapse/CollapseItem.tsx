import React, { useContext, useEffect, useRef, useState } from 'react';
import classNames from 'classnames';
import './index.scss';
import Icon from '../Icon'
import { CollapseContext } from './index'
import { CollapseItemAddProps, CollapseItemProps } from './type';

const CollapseItem: React.FC<CollapseItemProps> = (props) => {
  // 从 Collapse Context 注入全局属性
  const context = useContext(CollapseContext);
  const newProps: CollapseItemAddProps = context ? context.inject(props) : props;

  const {
    children,
    className,
    style,
    title,
    value,
    disabled = false,
    index,
    innerActive,
    updateInnerActive,
    iconPlacement = 'left',
    ...restProps
  } = newProps;

  const itemValue = value || index;
  const isActive = Array.isArray(innerActive) ? innerActive.includes(itemValue as string | number) : innerActive === itemValue;

  const handleClickHeader = () => {
    if (!disabled) {
      updateInnerActive?.(itemValue as string | number);
    }
  }

  const contentInnerRef = useRef<HTMLDivElement>(null)
  const [contentHeight, setContentHeight] = useState(0)
  useEffect(() => {
    const height = contentInnerRef.current?.getBoundingClientRect().height || 0
    setContentHeight(height + 16) // 加上下 padding

    const resizeObserver = new ResizeObserver(entries => {
      setContentHeight(entries[0].contentRect.height + 16)
    });
    resizeObserver.observe((contentInnerRef.current as any))
    return () => resizeObserver.disconnect()
  }, [])

  return (
    <div
      className={classNames(
        'i-collapse-item',
        disabled && 'i-collapse-item__disabled',
        iconPlacement === 'right' && 'i-collapse-item__icon-right',
        className
      )}
      style={{ ...style }}
      {...restProps}
    >
      <header
        className={classNames(
          'i-collapse-item__header',
          !isActive && 'i-collapse-item__header-fold'
        )}
        onClick={handleClickHeader}
        data-active={isActive ? 1 : 0}
      >
        <Icon
          name="ArrowCaretRight"
          style={{
            transform: isActive ? 'rotate(90deg)' : 'rotate(0deg)'
          }}
        />
        <span className="i-collapse-item__header-txt">
          {title}
        </span>
      </header>

      <section
        className={classNames(
          'i-collapse-item__content',
          isActive && 'i-collapse-item__content-unfold'
        )}
        style={{ height: isActive ? contentHeight : 0 }}
        data-active={isActive ? 1 : 0}
      >
        <div
          className="i-collapse-item__content-inner"
          ref={contentInnerRef}
        >
          {children}
        </div>
      </section>
    </div>
  )
}

export default CollapseItem;
