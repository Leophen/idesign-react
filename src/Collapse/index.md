---
nav:
  title: 组件
  path: /components
group:
  title: 数据展示组件
  order: 4
order: 8
---

# Collapse 折叠面板

用于折叠 / 展开内容区域的折叠面板组件。

## 基本用法

使用 `Collapse` 包裹菜单的每一项 `Collapse.Item`：

```tsx
import React, { useState } from 'react'
import { Collapse } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState([1])

  const handleChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <>
      <h4>无默认值</h4>
      <Collapse>
        <Collapse.Item title="折叠标题1">
          折叠内容 1111111111111111111111
        </Collapse.Item>
        <Collapse.Item title="折叠标题2">
          折叠内容 2222222222222222222222
        </Collapse.Item>
        <Collapse.Item title="折叠标题3">
          折叠内容 3333333333333333333333
        </Collapse.Item>
      </Collapse>
      <h4>有默认值</h4>
      <Collapse defaultActive={value}>
        <Collapse.Item title="折叠标题1">
          折叠内容 1111111111111111111111
        </Collapse.Item>
        <Collapse.Item title="折叠标题2">
          折叠内容 2222222222222222222222
        </Collapse.Item>
        <Collapse.Item title="折叠标题3">
          折叠内容 3333333333333333333333
        </Collapse.Item>
      </Collapse>
      <h4>有受控值</h4>
      <Collapse active={value}>
        <Collapse.Item title="折叠标题1">
          折叠内容 1111111111111111111111
        </Collapse.Item>
        <Collapse.Item title="折叠标题2">
          折叠内容 2222222222222222222222
        </Collapse.Item>
        <Collapse.Item title="折叠标题3">
          折叠内容 3333333333333333333333
        </Collapse.Item>
      </Collapse>
      <h4>一般用法</h4>
      <Collapse active={value} onChange={handleChange}>
        <Collapse.Item title="折叠标题1">
          折叠内容 1111111111111111111111
        </Collapse.Item>
        <Collapse.Item title="折叠标题2">
          折叠内容 2222222222222222222222
        </Collapse.Item>
        <Collapse.Item title="折叠标题3">
          折叠内容 3333333333333333333333
        </Collapse.Item>
      </Collapse>
    </>
  )
}

export default App
```

## 禁用状态

可通过 `disabled` 属性控制全局禁用或单项禁用：

```tsx
import React, { useState } from 'react'
import { Collapse } from 'idesign-react'

const App = () => {
  return (
    <>
      <h4>全局禁用</h4>
      <Collapse disabled>
        <Collapse.Item title="折叠标题1">
          折叠内容 1111111111111111111111
        </Collapse.Item>
        <Collapse.Item title="折叠标题2">
          折叠内容 2222222222222222222222
        </Collapse.Item>
        <Collapse.Item title="折叠标题3">
          折叠内容 3333333333333333333333
        </Collapse.Item>
      </Collapse>
      <h4>单项禁用</h4>
      <Collapse>
        <Collapse.Item title="折叠标题1">
          折叠内容 1111111111111111111111
        </Collapse.Item>
        <Collapse.Item title="折叠标题2" disabled>
          折叠内容 2222222222222222222222
        </Collapse.Item>
        <Collapse.Item title="折叠标题3">
          折叠内容 3333333333333333333333
        </Collapse.Item>
      </Collapse>
    </>
  )
}

export default App
```

## 手风琴模式

可通过 `accordion` 属性控制为手风琴模式：

```tsx
import React, { useState } from 'react'
import { Collapse } from 'idesign-react'

const App = () => {
  return (
    <Collapse accordion>
      <Collapse.Item title="折叠标题1">
        折叠内容 1111111111111111111111
      </Collapse.Item>
      <Collapse.Item title="折叠标题2">
        折叠内容 2222222222222222222222
      </Collapse.Item>
      <Collapse.Item title="折叠标题3">
        折叠内容 3333333333333333333333
      </Collapse.Item>
    </Collapse>
  )
}

export default App
```

## 默认展开全部

可通过 `expandAll` 属性控制默认展开所有折叠项：

```tsx
import React, { useState } from 'react'
import { Collapse } from 'idesign-react'

const App = () => {
  return (
    <Collapse expandAll>
      <Collapse.Item title="折叠标题1">
        折叠内容 1111111111111111111111
      </Collapse.Item>
      <Collapse.Item title="折叠标题2">
        折叠内容 2222222222222222222222
      </Collapse.Item>
      <Collapse.Item title="折叠标题3">
        折叠内容 3333333333333333333333
      </Collapse.Item>
    </Collapse>
  )
}

export default App
```

## 自定义图标位置

可通过 `iconPlacement` 属性控制展开图标位置：

```tsx
import React, { useState } from 'react'
import { Collapse } from 'idesign-react'

const App = () => {
  return (
    <>
      <h4>全局置左</h4>
      <Collapse iconPlacement='left'>
        <Collapse.Item title="折叠标题1">
          折叠内容 1111111111111111111111
        </Collapse.Item>
        <Collapse.Item title="折叠标题2">
          折叠内容 2222222222222222222222
        </Collapse.Item>
        <Collapse.Item title="折叠标题3">
          折叠内容 3333333333333333333333
        </Collapse.Item>
      </Collapse>
      <h4>全局置右</h4>
      <Collapse iconPlacement='right'>
        <Collapse.Item title="折叠标题1">
          折叠内容 1111111111111111111111
        </Collapse.Item>
        <Collapse.Item title="折叠标题2">
          折叠内容 2222222222222222222222
        </Collapse.Item>
        <Collapse.Item title="折叠标题3">
          折叠内容 3333333333333333333333
        </Collapse.Item>
      </Collapse>
      <h4>单项置右</h4>
      <Collapse>
        <Collapse.Item title="折叠标题1">
          折叠内容 1111111111111111111111
        </Collapse.Item>
        <Collapse.Item title="折叠标题2" iconPlacement='right'>
          折叠内容 2222222222222222222222
        </Collapse.Item>
        <Collapse.Item title="折叠标题3">
          折叠内容 3333333333333333333333
        </Collapse.Item>
      </Collapse>
    </>
  )
}

export default App
```

## 隐藏所有边框

可通过 `hideBorder` 属性隐藏所有边框：

```tsx
import React, { useState } from 'react'
import { Collapse } from 'idesign-react'

const App = () => {
  return (
    <Collapse hideBorder>
      <Collapse.Item title="折叠标题1">
        折叠内容 1111111111111111111111
      </Collapse.Item>
      <Collapse.Item title="折叠标题2">
        折叠内容 2222222222222222222222
      </Collapse.Item>
      <Collapse.Item title="折叠标题3">
        折叠内容 3333333333333333333333
      </Collapse.Item>
    </Collapse>
  )
}

export default App
```

## 嵌套折叠面板

可在 `Collapse.Item` 中多层嵌套 `Collapse` 折叠面板：

```tsx
import React, { useState } from 'react'
import { Collapse } from 'idesign-react'

const App = () => {
  return (
    <Collapse>
      <Collapse.Item title="折叠标题1">
        <Collapse hideBorder>
          <Collapse.Item title="折叠标题A">
            折叠内容 AAAAAAAAAAAAAAAAAAAAAA
          </Collapse.Item>
          <Collapse.Item title="折叠标题B">
            折叠内容 BBBBBBBBBBBBBBBBBBBBBB
          </Collapse.Item>
          <Collapse.Item title="折叠标题C">
            折叠内容 CCCCCCCCCCCCCCCCCCCCCC
          </Collapse.Item>
        </Collapse>
      </Collapse.Item>
      <Collapse.Item title="折叠标题2">
        折叠内容 2222222222222222222222
      </Collapse.Item>
      <Collapse.Item title="折叠标题3">
        折叠内容 3333333333333333333333
      </Collapse.Item>
    </Collapse>
  )
}

export default App
```

## 无缩进模式

可通过 `noIndent` 属性隐藏所有边框：

```tsx
import React, { useState } from 'react'
import { Collapse } from 'idesign-react'

const App = () => {
  return (
    <Collapse noIndent>
      <Collapse.Item title="折叠标题1">
        <Collapse hideBorder>
          <Collapse.Item title="折叠标题A">
            折叠内容 AAAAAAAAAAAAAAAAAAAAAA
          </Collapse.Item>
          <Collapse.Item title="折叠标题B">
            折叠内容 BBBBBBBBBBBBBBBBBBBBBB
          </Collapse.Item>
          <Collapse.Item title="折叠标题C">
            折叠内容 CCCCCCCCCCCCCCCCCCCCCC
          </Collapse.Item>
        </Collapse>
      </Collapse.Item>
      <Collapse.Item title="折叠标题2">
        折叠内容 2222222222222222222222
      </Collapse.Item>
      <Collapse.Item title="折叠标题3">
        折叠内容 3333333333333333333333
      </Collapse.Item>
    </Collapse>
  )
}

export default App
```

## Collapse API

<API hideTitle />

## CollapseItem API

| 属性          | 说明           | 类型              | 默认值  |
| ------------- | -------------- | ----------------- | ------- |
| children      | 按钮内容       | `ReactNode`       | `--`    |
| className     | 类名           | `string`          | `--`    |
| style         | 自定义样式     | `CSSProperties`   | `--`    |
| title         | 折叠项标题     | `ReactNode`       | `--`    |
| value         | 折叠项唯一标识 | `string〡number`  | `--`    |
| disabled      | 禁用单折叠项   | `boolean`         | `false` |
| iconPlacement | 自定义图标位置 | `'left'〡'right'` | `left`  |
