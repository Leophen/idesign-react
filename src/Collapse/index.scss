.i-collapse {
  box-sizing: border-box;
  overflow: hidden;
  color: var(--i-font-2);
  font-size: 14px;
  border: 1px solid var(--i-border);
  border-radius: 6px;
  outline: none;
  &.i-collapse__hide-border {
    border: none;
    .i-collapse-item__active,
    .i-collapse-item__header,
    .i-collapse-item__header-fold,
    .i-collapse-item__content,
    .i-collapse-item__content-unfold {
      border: none;
    }
    .i-collapse-item:last-of-type {
      .i-collapse-item__header-fold {
        border: none;
      }
    }
  }
  &.i-collapse__no-indent {
    .i-collapse-item__content {
      padding-left: 12px;
      .i-collapse-item__content {
        padding-left: 0;
      }
    }
  }
}

.i-collapse-item {
  box-sizing: border-box;
  border-bottom: none;
  outline: none;
  &:last-of-type {
    .i-collapse-item__content-unfold {
      border-bottom: none;
    }
    .i-collapse-item__header-fold {
      border-bottom: 1px solid transparent;
    }
  }
  &.i-collapse-item__icon-right {
    .i-collapse-item__header {
      flex-direction: row-reverse;
    }
    .i-collapse-item__header-txt {
      flex: 1;
    }
    .i-collapse-item__content {
      padding-left: 24px;
    }
  }
  &.i-collapse-item__active {
    border-bottom: 1px solid var(--i-border);
  }
  &.i-collapse-item__disabled {
    .i-collapse-item__header {
      color: var(--i-font-disabled);
      cursor: not-allowed;
      .icon-ArrowCaretRight {
        color: var(--i-font-disabled);
      }
    }
  }
}

.i-collapse-item__header {
  display: flex;
  gap: 6px;
  align-items: center;
  box-sizing: border-box;
  height: 40px;
  padding: 0 12px;
  border-bottom: 1px solid var(--i-border);
  cursor: pointer;
  transition: all 0.2s ease-in;
  .icon-ArrowCaretRight {
    color: var(--i-font-2);
    transition: all 0.2s ease-out;
  }
}

.i-collapse-item__content {
  box-sizing: border-box;
  padding: 0 12px 0 34px;
  overflow: hidden;
  background: var(--i-bg);
  visibility: hidden;
  transition: all 0.2s ease-out;
  &.i-collapse-item__content-unfold {
    padding: 8px 12px 8px 34px;
    border-bottom: 1px solid var(--i-border);
    visibility: visible;
  }
  .i-collapse-item__header {
    padding: 0;
  }
  .i-collapse-item__content {
    padding-left: 22px;
  }
}
