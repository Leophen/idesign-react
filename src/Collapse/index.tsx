import React from 'react';
import classNames from 'classnames';
import './index.scss';
import useDefault from '../hooks/useDefault';
import CollapseItem from './CollapseItem';
import { CollapseContextValue, CollapseItemProps, CollapseProps, CollapseValueType } from './type';

export const CollapseContext = React.createContext<CollapseContextValue>(null as any);

const Collapse: React.FC<CollapseProps> & { Item: React.ElementType } = (props) => {
  let defaultVal: CollapseValueType = []
  React.Children.map(props.children, (child, index) => {
    !props.defaultActive && props.expandAll && (defaultVal.push((child as any).props.value || index))
  })

  const {
    children = '',
    className,
    style,
    active,
    defaultActive = defaultVal,
    accordion = false,
    disabled = false,
    expandAll = false,
    iconPlacement = 'left',
    hideBorder = false,
    noIndent = false,
    onChange,
    ...restProps
  } = props;

  const [innerActive, setInnerActive] = useDefault(active, defaultActive, onChange);

  // 更新展开项
  const updateInnerActive = (value: string | number) => {
    let newValue: CollapseValueType = [].concat(innerActive as any || []);
    const index = newValue.indexOf(value);
    if (index >= 0) {
      newValue.splice(index, 1);
    } else if (accordion) {
      newValue = [value];
    } else {
      newValue.push(value);
    }
    setInnerActive([...newValue]);
  };

  // 注入每一项的 context
  const context: CollapseContextValue = {
    // 将 props 注入每一项子节点的方法
    inject: (singleCollapseProps: CollapseItemProps) => {
      return {
        innerActive,
        disabled,
        updateInnerActive,
        iconPlacement,
        ...singleCollapseProps,
      };
    },
  };

  return (
    <CollapseContext.Provider value={context}>
      <div
        className={classNames(
          'i-collapse',
          hideBorder && 'i-collapse__hide-border',
          noIndent && 'i-collapse__no-indent',
          className
        )}
        style={{ ...style }}
        {...restProps}
      >
        {React.Children.map(children, (child, index) => {
          if (!React.isValidElement(child)) {
            return null;
          }
          const childProps = {
            index,
            ...child.props,
          };
          return React.cloneElement(child, childProps);
        })}
      </div>
    </CollapseContext.Provider>
  );
};

Collapse.Item = CollapseItem;

CollapseItem.displayName = 'CollapseItem';
Collapse.displayName = 'Collapse';

export default Collapse;
