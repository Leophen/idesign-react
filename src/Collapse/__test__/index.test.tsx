import { render } from '@testing-library/react';
import Collapse from '../index';

describe('Collapse 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(
      <Collapse>
        <Collapse.Item title="折叠标题1">
          折叠内容 1
        </Collapse.Item>
        <Collapse.Item title="折叠标题2">
          折叠内容 2
        </Collapse.Item>
        <Collapse.Item title="折叠标题3">
          折叠内容 3
        </Collapse.Item>
      </Collapse>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

