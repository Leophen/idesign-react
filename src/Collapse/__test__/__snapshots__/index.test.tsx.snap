// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-collapse"
  >
    <div
      class="i-collapse-item"
    >
      <header
        class="i-collapse-item__header i-collapse-item__header-fold"
        data-active="0"
      >
        <i
          class="i-icon icon-ArrowCaretRight"
          style="transform: rotate(0deg);"
        />
        <span
          class="i-collapse-item__header-txt"
        >
          折叠标题1
        </span>
      </header>
      <section
        class="i-collapse-item__content"
        data-active="0"
        style="height: 0px;"
      >
        <div
          class="i-collapse-item__content-inner"
        >
          折叠内容 1
        </div>
      </section>
    </div>
    <div
      class="i-collapse-item"
    >
      <header
        class="i-collapse-item__header i-collapse-item__header-fold"
        data-active="0"
      >
        <i
          class="i-icon icon-ArrowCaretRight"
          style="transform: rotate(0deg);"
        />
        <span
          class="i-collapse-item__header-txt"
        >
          折叠标题2
        </span>
      </header>
      <section
        class="i-collapse-item__content"
        data-active="0"
        style="height: 0px;"
      >
        <div
          class="i-collapse-item__content-inner"
        >
          折叠内容 2
        </div>
      </section>
    </div>
    <div
      class="i-collapse-item"
    >
      <header
        class="i-collapse-item__header i-collapse-item__header-fold"
        data-active="0"
      >
        <i
          class="i-icon icon-ArrowCaretRight"
          style="transform: rotate(0deg);"
        />
        <span
          class="i-collapse-item__header-txt"
        >
          折叠标题3
        </span>
      </header>
      <section
        class="i-collapse-item__content"
        data-active="0"
        style="height: 0px;"
      >
        <div
          class="i-collapse-item__content-inner"
        >
          折叠内容 3
        </div>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
